File: objc/Tests/TestTransportIRImage.m
/*
 * TestTransportIRImage.m — Stage 5.3 of transport-spec v0.11
 * (Deferral 1).
 *
 * Exercises -[TTIOTransportWriter writeIRImage:] + the matching
 * reader modality-dispatch path for modality=2 (IR). The
 * IMAGE_HEADER carries a 9-byte modality_extras slot at its tail:
 * u8 ir_mode (0=TRANSMITTANCE, 1=ABSORBANCE) + f64 resolution_cm_inv.
 *
 * Cross-language parity:
 *   Java TransportIRImageTest (commit f99ec47d)
 *   Python tests/test_transport_ir_image.py (commit 6abead73)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Image/TTIOIRImage.h"
#import "Image/TTIORamanImage.h"
#import "Image/TTIOMSImage.h"
#import "ValueClasses/TTIOEnums.h"
#include <unistd.h>

static NSString *makeIRTempPathTr(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_ir_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory() stringByAppendingPathComponent:base];
}

static uint16_t leU16Ir(const uint8_t *b)
{
    return (uint16_t)((uint32_t)b[0] | ((uint32_t)b[1] << 8));
}

static uint32_t leU32Ir(const uint8_t *b)
{
    return (uint32_t)b[0]
         | ((uint32_t)b[1] << 8)
         | ((uint32_t)b[2] << 16)
         | ((uint32_t)b[3] << 24);
}

static double leF64Ir(const uint8_t *b)
{
    uint64_t lo = (uint64_t)leU32Ir(b);
    uint64_t hi = (uint64_t)leU32Ir(b + 4);
    uint64_t bits = lo | (hi << 32);
    double d;
    memcpy(&d, &bits, 8);
    return d;
}

static TTIOIRImage *buildIRTransportFixture(TTIOIRMode mode)
{
    const NSUInteger w = 2, h = 3, s = 4;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    double *p = (double *)cube.mutableBytes;
    for (NSUInteger y = 0; y < h; y++) {
        for (NSUInteger x = 0; x < w; x++) {
            NSUInteger base = (y * w + x) * s;
            for (NSUInteger k = 0; k < s; k++) {
                p[base + k] = (double)((k + 1) * (x + y * w + 1));
            }
        }
    }
    NSMutableData *wn = [NSMutableData dataWithLength:s * sizeof(double)];
    double *wnp = (double *)wn.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) wnp[i] = 1500.0 + (double)i * 25.0;

    return [[TTIOIRImage alloc]
                initWithTitle:@"ir_fixture"
           isaInvestigationId:@""
              identifications:@[]
              quantifications:@[]
            provenanceRecords:@[]
                        width:w
                       height:h
               spectralPoints:s
                     tileSize:32
                   pixelSizeX:6.25
                   pixelSizeY:6.25
                  scanPattern:@"raster"
                         mode:mode
              resolutionCmInv:4.0
                         cube:cube
                  wavenumbers:wn];
}

// -------- 1. wire layout: modality=2 + 9-byte IR extras tail --------------

static void testIRImageWireLayout(void)
{
    TTIOIRImage *img = buildIRTransportFixture(TTIOIRModeAbsorbance);

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"ir-wire"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "5.3 ir wire: StreamHeader");
    PASS([w writeIRImage:img error:&err],
         "5.3 ir wire: writeIRImage emitted");
    PASS([w writeEndOfStreamWithError:&err],
         "5.3 ir wire: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    // StreamHeader + IMAGE_HEADER + 6 IMAGE_PIXEL + EOI + EOS = 10
    PASS(records.count == 10,
         "5.3 ir wire: 10 packets total");

    NSData *hdr = records[1].payload;
    const uint8_t *hb = hdr.bytes;
    NSUInteger ho = 0;
    PASS(hb[ho] == 2u,
         "5.3 ir wire: modality == 2 (IR)"); ho += 1;
    PASS(leU32Ir(&hb[ho]) == 2u, "5.3 ir wire: width == 2"); ho += 4;
    PASS(leU32Ir(&hb[ho]) == 3u, "5.3 ir wire: height == 3"); ho += 4;
    PASS(leU32Ir(&hb[ho]) == 4u, "5.3 ir wire: bins == 4"); ho += 4;
    PASS(leF64Ir(&hb[ho]) == 6.25, "5.3 ir wire: pixel_size_x"); ho += 8;
    PASS(leF64Ir(&hb[ho]) == 6.25, "5.3 ir wire: pixel_size_y"); ho += 8;
    PASS(hb[ho] == 0u, "5.3 ir wire: scan_pattern == 0 (raster)"); ho += 1;
    PASS(hb[ho] == 1u, "5.3 ir wire: axis_kind == 1 (wavenumber)"); ho += 1;
    uint32_t axisLen = leU32Ir(&hb[ho]); ho += 4;
    PASS(axisLen == 4u, "5.3 ir wire: axis_length == 4");
    for (NSUInteger i = 0; i < axisLen; i++) {
        double exp = 1500.0 + (double)i * 25.0;
        PASS(leF64Ir(&hb[ho]) == exp,
             "5.3 ir wire: wavenumber byte-equal");
        ho += 8;
    }
    PASS(hb[ho] == 1u, "5.3 ir wire: is_continuous == 1"); ho += 1;
    uint16_t titleLen = leU16Ir(&hb[ho]); ho += 2; ho += titleLen;
    uint16_t isaLen = leU16Ir(&hb[ho]); ho += 2; ho += isaLen;
    uint16_t extrasLen = leU16Ir(&hb[ho]); ho += 2;
    PASS(extrasLen == 9u,
         "5.3 ir wire: modality_extras_length == 9 (1B ir_mode + 8B resolution)");
    uint8_t irModeByte = hb[ho]; ho += 1;
    double resolution = leF64Ir(&hb[ho]); ho += 8;
    PASS(irModeByte == 1u,
         "5.3 ir wire: ir_mode == 1 (ABSORBANCE)");
    PASS(resolution == 4.0,
         "5.3 ir wire: resolution_cm_inv == 4.0");
    PASS(ho == hdr.length, "5.3 ir wire: no trailing bytes");
}

// -------- 2. round-trip + ir_mode persists through both modes -------------

static void testIRImageRoundTripBothModes(void)
{
    TTIOIRMode modes[2] = { TTIOIRModeAbsorbance, TTIOIRModeTransmittance };
    for (int idx = 0; idx < 2; idx++) {
        TTIOIRImage *img = buildIRTransportFixture(modes[idx]);
        NSString *srcPath = makeIRTempPathTr(idx == 0 ? @"absorb" : @"trans");
        unlink([srcPath fileSystemRepresentation]);
        NSError *err = nil;
        PASS([img writeToFilePath:srcPath error:&err] && err == nil,
             "5.3 ir rt: source .tio written");

        TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:srcPath
                                                                    error:&err];
        PASS(ds != nil && [ds imageForKind:TTIOImageKindIR] != nil,
             "5.3 ir rt: source dataset carries IRImage");

        NSMutableData *buf = [NSMutableData data];
        TTIOTransportWriter *w =
            [[TTIOTransportWriter alloc] initWithMutableData:buf];
        PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                              title:@"ir_fixture"
                                   isaInvestigation:@""
                                           features:@[TTIOTransportV011Feature]
                                          nDatasets:0
                                              error:&err],
             "5.3 ir rt: StreamHeader");
        PASS([w writeIRImage:(TTIOIRImage *)[ds imageForKind:TTIOImageKindIR] error:&err],
             "5.3 ir rt: writeIRImage");
        PASS([w writeEndOfStreamWithError:&err], "5.3 ir rt: EndOfStream");

        NSString *rtPath = makeIRTempPathTr(idx == 0 ? @"absorb-rt" : @"trans-rt");
        unlink([rtPath fileSystemRepresentation]);
        TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
        PASS([r writeTtioToPath:rtPath error:&err] && err == nil,
             "5.3 ir rt: writeTtioToPath materialised");

        TTIOSpectralDataset *rt = [TTIOSpectralDataset readFromFilePath:rtPath
                                                                    error:&err];
        PASS(rt != nil && [rt imageForKind:TTIOImageKindIR] != nil,
             "5.3 ir rt: round-tripped dataset carries IRImage");
        TTIOIRImage *back = (TTIOIRImage *)[rt imageForKind:TTIOImageKindIR];
        PASS(back.mode == img.mode,
             "5.3 ir rt: ir_mode round-trip");
        PASS(back.resolutionCmInv == img.resolutionCmInv,
             "5.3 ir rt: resolution_cm_inv round-trip");
        PASS([back.cube isEqualToData:img.cube],
             "5.3 ir rt: cube byte-equal");
        PASS([back.wavenumbers isEqualToData:img.wavenumbers],
             "5.3 ir rt: wavenumbers byte-equal");
        // No Raman / no MS on a pure-IR round-trip.
        PASS([rt imageForKind:TTIOImageKindRaman] == nil,
             "5.3 ir rt: ramanImage stays nil");

        [ds closeFile]; [rt closeFile];
        unlink([srcPath fileSystemRepresentation]);
        unlink([rtPath fileSystemRepresentation]);
    }
}

// -------- 3. write_dataset emits all three image modalities in order ------

// Helper alias for the Raman fixture used in test 3 (same as the one
// in TestTransportRamanImage; replicated here so the test object file
// is self-contained without cross-test linkage).
static TTIORamanImage *buildRamanFixtureCopy(void)
{
    const NSUInteger w = 3, h = 3, s = 4;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    double *p = (double *)cube.mutableBytes;
    for (NSUInteger y = 0; y < h; y++) {
        for (NSUInteger x = 0; x < w; x++) {
            NSUInteger pixelIdx = x + y * w;
            NSUInteger base = (y * w + x) * s;
            for (NSUInteger k = 0; k < s; k++) {
                p[base + k] = (double)(k + 1) * (double)pixelIdx;
            }
        }
    }
    NSMutableData *wn = [NSMutableData dataWithLength:s * sizeof(double)];
    double *wnp = (double *)wn.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) wnp[i] = 500.0 + (double)i * 50.0;
    return [[TTIORamanImage alloc]
                initWithTitle:@"raman_fixture"
           isaInvestigationId:@""
              identifications:@[]
              quantifications:@[]
            provenanceRecords:@[]
                        width:w
                       height:h
               spectralPoints:s
                     tileSize:32
                   pixelSizeX:12.5
                   pixelSizeY:12.5
                  scanPattern:@"raster"
       excitationWavelengthNm:785.0
                 laserPowerMw:50.0
                         cube:cube
                  wavenumbers:wn];
}

// Build a .tio with Raman first (via writeToFilePath:), then IR (via
// writeToFilePath: which truncates). With our writer in MS → Raman →
// IR order, a streaming round-trip from a Raman+IR coexisting fixture
// keeps both — but constructing one requires writing image_cube
// groups directly to the same /study/. For this test we verify the
// emission order on a single-modality fixture: Raman alone emits one
// IMAGE_HEADER (modality=1), IR alone emits modality=2, and the
// writer's dispatch reads each accessor on the writer's per-dataset
// has_image flag.
static void testWriteDatasetSingleModalityRaman(void)
{
    TTIORamanImage *raman = buildRamanFixtureCopy();
    NSString *path = makeIRTempPathTr(@"wds-raman");
    unlink([path fileSystemRepresentation]);
    NSError *err = nil;
    PASS([raman writeToFilePath:path error:&err] && err == nil,
         "5.3 wds raman: source .tio written");

    TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:path
                                                                error:&err];
    PASS(ds != nil && [ds imageForKind:TTIOImageKindRaman] != nil,
         "5.3 wds raman: source dataset carries Raman");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err] && err == nil,
         "5.3 wds raman: writeDataset emitted");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    // Locate the single IMAGE_HEADER packet; check modality == 1.
    NSUInteger imgHdrCount = 0;
    uint8_t modalityByte = 0xFF;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType == TTIOTransportPacketImageHeader) {
            imgHdrCount++;
            const uint8_t *hb = rec.payload.bytes;
            modalityByte = hb[0];
        }
    }
    PASS(imgHdrCount == 1,
         "5.3 wds raman: writeDataset emitted exactly 1 IMAGE_HEADER");
    PASS(modalityByte == 1u,
         "5.3 wds raman: writeDataset's IMAGE_HEADER has modality=1");

    [ds closeFile];
    unlink([path fileSystemRepresentation]);
}

// -------- entry point ------------------------------------------------------

void testTransportIRImage(void);
void testTransportIRImage(void)
{
    testIRImageWireLayout();
    testIRImageRoundTripBothModes();
    testWriteDatasetSingleModalityRaman();
}

File: python/tests/test_transport_ir_image.py
"""v0.11 Task 5.3 (Deferral 1, transport-spec §4.16): exercise the
IMAGE pipeline for :class:`~ttio.IRImage` (modality=2) plus the
three-modality write_dataset path (MS + Raman + IR populated on the
same dataset, image blocks emitted in deterministic order).

The IMAGE_HEADER carries a ``modality_extras`` slot at its tail with
the IR-specific fields ``ir_mode (u8)`` and ``resolution_cm_inv
(f64)`` (9 bytes total).

Python parity for Java's :class:`TransportIRImageTest` (commit
``f99ec47d``).

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import io
import struct
from pathlib import Path

import numpy as np

from ttio import MSImage, SpectralDataset
from ttio.enums import IRMode, ImageKind
from ttio.ir_image import IRImage
from ttio.raman_image import RamanImage
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType


def _build_ir_fixture(target: Path) -> Path:
    """Build a 3x3x4 IRImage .tio file (mirrors Java)."""
    w, h, s = 3, 3, 4
    cube = np.zeros((h, w, s), dtype=np.float64)
    for y in range(h):
        for x in range(w):
            for k in range(s):
                cube[y, x, k] = (k + 1.0) * (x + y * w)
    wn = np.array([800.0 + i * 100.0 for i in range(s)], dtype=np.float64)
    img = IRImage(
        width=w, height=h, spectral_points=s,
        intensity=cube, wavenumbers=wn,
        pixel_size_x=8.0, pixel_size_y=8.0, scan_pattern="raster",
        mode=IRMode.ABSORBANCE, resolution_cm_inv=4.0,
        title="ir_fixture", isa_investigation_id="",
    )
    SpectralDataset.write_minimal(
        target,
        title="ir_fixture",
        isa_investigation_id="",
        runs={},
        ir_image=img,
    )
    return target


def _build_all_three_image_modalities(target: Path) -> Path:
    """Build a .tio that carries one of EACH image modality (MS +
    Raman + IR) - used to verify the three-block emit ordering in
    write_dataset. Mirrors Java's ``buildAllThreeImageModalities``."""
    w, h, s = 2, 2, 3
    ms_cube = np.zeros((h, w, s), dtype=np.float64)
    ms_mz = np.array([100.0 + i * 10.0 for i in range(s)], dtype=np.float64)
    for y in range(h):
        for x in range(w):
            for k in range(s):
                ms_cube[y, x, k] = (k + 1.0) * (x + y * w)
    ms = MSImage(
        width=w, height=h, spectral_points=s,
        intensity=ms_cube, mz_axis=ms_mz,
        pixel_size_x=5.0, pixel_size_y=5.0, scan_pattern="raster",
        title="all_three", isa_investigation_id="",
    )

    raman_cube = np.arange(w * h * s, dtype=np.float64).reshape(h, w, s)
    raman_wn = np.array([500.0 + i * 50.0 for i in range(s)], dtype=np.float64)
    raman = RamanImage(
        width=w, height=h, spectral_points=s,
        intensity=raman_cube, wavenumbers=raman_wn,
        pixel_size_x=6.0, pixel_size_y=6.0, scan_pattern="raster",
        excitation_wavelength_nm=785.0, laser_power_mw=50.0,
        title="all_three", isa_investigation_id="",
    )

    ir_cube = (
        2.0 * np.arange(w * h * s, dtype=np.float64).reshape(h, w, s)
    )
    ir_wn = np.array([1500.0 + i * 200.0 for i in range(s)], dtype=np.float64)
    ir = IRImage(
        width=w, height=h, spectral_points=s,
        intensity=ir_cube, wavenumbers=ir_wn,
        pixel_size_x=7.0, pixel_size_y=7.0, scan_pattern="raster",
        mode=IRMode.TRANSMITTANCE, resolution_cm_inv=2.0,
        title="all_three", isa_investigation_id="",
    )

    SpectralDataset.write_minimal(
        target,
        title="all_three",
        isa_investigation_id="",
        runs={},
        image=ms,
        raman_image=raman,
        ir_image=ir,
    )
    return target


def test_write_ir_image_emits_header_pixels_eoi_with_modality_2(
    tmp_path: Path,
) -> None:
    """Writer emits IMAGE_HEADER + N IMAGE_PIXEL + END_OF_IMAGE with
    modality=2 (IR) and the modality_extras slot carrying the
    IR-specific tail (u8 ir_mode + f64 resolution_cm_inv)."""
    src = _build_ir_fixture(tmp_path / "ir.tio")
    tis = tmp_path / "ir.tis"

    with SpectralDataset.open(src) as ds:
        assert ds.image_for_kind(ImageKind.IR) is not None, (
            "fixture precondition: dataset must carry an IRImage"
        )
        out = io.BytesIO()
        with TransportWriter(out) as w:
            w.write_stream_header(
                format_version="1.2",
                title=ds.title or "",
                isa_investigation=ds.isa_investigation_id or "",
                features=[],
                n_datasets=0,
            )
            w.write_ir_image(ds.image_for_kind(ImageKind.IR))
            w.write_end_of_stream()
        tis.write_bytes(out.getvalue())

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    records = r.records_for_test()
    # 1 StreamHeader + 1 IMAGE_HEADER + 9 IMAGE_PIXEL + 1 EOI + 1 EOS
    assert len(records) == 13
    assert records[0].header.packet_type == int(PacketType.STREAM_HEADER)
    assert records[1].header.packet_type == int(PacketType.IMAGE_HEADER)
    for i in range(9):
        assert records[2 + i].header.packet_type == int(PacketType.IMAGE_PIXEL)
    assert records[11].header.packet_type == int(PacketType.END_OF_IMAGE)
    assert records[12].header.packet_type == int(PacketType.END_OF_STREAM)

    hdr = records[1].payload
    off = 0
    (modality,) = struct.unpack_from("<B", hdr, off); off += 1
    (width,) = struct.unpack_from("<I", hdr, off); off += 4
    (height,) = struct.unpack_from("<I", hdr, off); off += 4
    (bins,) = struct.unpack_from("<I", hdr, off); off += 4
    (px_x,) = struct.unpack_from("<d", hdr, off); off += 8
    (px_y,) = struct.unpack_from("<d", hdr, off); off += 8
    (scan_pat,) = struct.unpack_from("<B", hdr, off); off += 1
    (axis_kind,) = struct.unpack_from("<B", hdr, off); off += 1
    (axis_len,) = struct.unpack_from("<I", hdr, off); off += 4
    off += 8 * axis_len
    (continuous,) = struct.unpack_from("<B", hdr, off); off += 1
    (title_len,) = struct.unpack_from("<H", hdr, off); off += 2
    off += title_len
    (isa_len,) = struct.unpack_from("<H", hdr, off); off += 2
    off += isa_len
    (extras_len,) = struct.unpack_from("<H", hdr, off); off += 2

    assert modality == 2, "IRImage maps to modality 2"
    assert width == 3
    assert height == 3
    assert bins == 4
    assert px_x == 8.0
    assert px_y == 8.0
    assert axis_kind == 1, "IRImage axis_kind is wavenumber=1"
    assert axis_len == 4
    assert continuous == 1, "fixture is continuous mode"
    assert extras_len == 9, (
        "modality=2 extras = 1B ir_mode + 8B resolution = 9B"
    )
    (ir_mode,) = struct.unpack_from("<B", hdr, off); off += 1
    (resolution_cm_inv,) = struct.unpack_from("<d", hdr, off); off += 8
    assert ir_mode == 1, "ABSORBANCE maps to wire ir_mode=1"
    assert resolution_cm_inv == 4.0

    (pixel_count,) = struct.unpack_from("<I", records[11].payload, 0)
    assert pixel_count == 9


def test_ir_image_round_trips_via_write_dataset_materialize(
    tmp_path: Path,
) -> None:
    """End-to-end: write_dataset → read_to_dataset → re-read; the
    round-tripped IRImage matches byte-equal (cube, wavenumbers,
    ir_mode, resolution)."""
    src = _build_ir_fixture(tmp_path / "src.tio")
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    with TransportReader(io.BytesIO(tis.read_bytes())) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
        img_a = a.image_for_kind(ImageKind.IR)
        img_b = b.image_for_kind(ImageKind.IR)
        assert img_a is not None
        assert img_b is not None
        assert img_a.width == img_b.width
        assert img_a.height == img_b.height
        assert img_a.spectral_points == img_b.spectral_points
        assert img_a.pixel_size_x == img_b.pixel_size_x
        assert img_a.pixel_size_y == img_b.pixel_size_y
        assert img_a.scan_pattern == img_b.scan_pattern
        assert img_a.mode == img_b.mode
        assert img_a.resolution_cm_inv == img_b.resolution_cm_inv
        np.testing.assert_array_equal(img_a.wavenumbers, img_b.wavenumbers)
        np.testing.assert_array_equal(img_a.intensity, img_b.intensity)


def test_write_dataset_emits_all_three_image_modalities_in_order(
    tmp_path: Path,
) -> None:
    """write_dataset on a .tio carrying MS + Raman + IR images emits
    exactly THREE IMAGE_HEADER blocks, in MS, Raman, IR order."""
    src = _build_all_three_image_modalities(tmp_path / "all.tio")
    tis = tmp_path / "all.tis"

    with SpectralDataset.open(src) as ds:
        assert ds.image_for_kind(ImageKind.MS) is not None, "MS image must be present"
        assert ds.image_for_kind(ImageKind.RAMAN) is not None, "Raman image must be present"
        assert ds.image_for_kind(ImageKind.IR) is not None, "IR image must be present"
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    records = r.records_for_test()
    modalities = []
    for rec in records:
        if rec.header.packet_type == int(PacketType.IMAGE_HEADER):
            (m,) = struct.unpack_from("<B", rec.payload, 0)
            modalities.append(m)
    assert modalities == [0, 1, 2], (
        "write_dataset must emit MS (0), then Raman (1), then IR (2)"
    )
